### DISCUSSION TITLE
Need to check if user has a permission via one of their roles

### DISCUSSION THREAD

**Discussion by salamwaddah** [spatie/laravel-permission#2330] [Q&A] [answered]
I am trying to prevent attaching a permission directly to a user if they have said permission via one of their roles and `hasPermissionViaRole` method in `HasPermissions` trait does exactly that but it's `proctected`. My `User` models uses `HasRoles` trait which uses `HasPermissions` trait.

Code in controller

```php
if ($user->hasPermissionViaRole($permission)) {
  // do not allow attaching permission directly
}
```
Gives this error 
> BadMethodCallException: Call to undefined method User::hasPermissionViaRole()

### Question
Any reason `hasPermissionViaRole` method is protected? Similar discussions related to this method suggested I should just override it in the User model.

<img width="654" alt="image" src="https://user-images.githubusercontent.com/26627088/218656704-3e812532-4a11-4971-a6b2-5e2e6f80245d.png">

**Comment by erikn69** [comment, marked as answer]
just override it in the User model or make a PR to change it

**Comment by drbyte** [comment]
Also, `hasPermissionTo()` checks both permissions-via-roles and direct permissions.
So, you could just call that to see whether the user already has the permission.

**Reply by salamwaddah** [reply]
I understand that, I wanted to explicitly check for inherited permissions. I choose to override the protected method.
